rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+# RFC 9811 - 2021-11-12 - VRL enum VM (VenuM)
+
+This RFC proposes implementing a enum VM. VRL will be compiled to a list of
+instructions and executed by this VM, with the aim to significantly improve the
+performance of executing VRL programs.
+
+## Context
+
+This RFC is a follow on from
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Scope
+
+### In scope
+
+This RFC is purely about developing a enum VM for VRL. It will discuss the
+risks involved in running VRL with a VM and the measures we will take to
+mitigate those risks.
+
+### Out of scope
```

Comment:
It came to mind that I think it would be good to add "user-facing breaking changes" as out of scope for this RFC. That is, whatever implementation we end up with, there shouldn't be any difference to the user, other than a faster runtime.

If we do want to do optimizations that require a breaking change, I suggest pulling those out into a separate issue, and discuss them separately, so that we don't block any work listed in this RFC.

Reply:
Yes! It is very much the plan as a part of the testing to be able to run the current implementation alongside the VM to make sure the output doesn't change in any way.

Reply:
Yes, to add to this, there should be _no_ user facing changes other than VRL is faster. I've talked with Stephen offline a bit about this but I'd like to see us have property tests that assert the equivalent operation of the tree walker and the enum VM. That'll allow us to be sure any future optimization passes don't introduce gross regressions.